src/software/ai/hl/stp/tactic/receiver/receiver_fsm.h

```diff
@@ -0,0 +1,296 @@
+        }
+        Ray shot(closest_ball_pos, best_shot_target - closest_ball_pos);
+
+        Angle ideal_orientation      = getOneTimeShotDirection(shot, ball);
+        Vector ideal_orientation_vec = Vector::createFromAngle(ideal_orientation);
+
+        // The best position is determined such that the robot stays in the ideal
+        // orientation, but moves the shortest distance possible to put its contact
+        // point in the ball's path.
+        Point ideal_position =
+            closest_ball_pos - ideal_orientation_vec.normalize(dist_to_ball_in_dribbler);
+
+        return Shot(ideal_position, ideal_orientation);
+    }
+
+    /*
+     * Finds a shot that is greater than MIN_SHOT_NET_PERCENT_OPEN and
+     * respects MAX_DEFLECTION_FOR_ONE_TOUCH_SHOT for the highest chance
+     * of scoring with a one-touch shot. If neither of those are true, return a nullopt
+     *
+     * @param world The world to find a feasible shot on
+     * @param assigned_robot The robot that will be performing the one-touch
+     */
+    static std::optional<Shot> findFeasibleShot(const World& world,
+                                                const Robot& assigned_robot)
+    {
+        // Check if we can shoot on the enemy goal from the receiver position
+        std::optional<Shot> best_shot_opt = calcBestShotOnGoal(
+            world.field(), world.friendlyTeam(), world.enemyTeam(),
+            assigned_robot.position(), TeamType::ENEMY, {assigned_robot});
+
+        // Vector from the ball to the robot
+        Vector robot_to_ball = world.ball().position() - assigned_robot.position();
+
+        // The angle the robot will have to deflect the ball to shoot
+        Angle abs_angle_between_pass_and_shot_vectors;
+        // The percentage of open net the robot would shoot on
+        double net_percent_open;
+        if (best_shot_opt)
+        {
+            Vector robot_to_shot_target =
+                best_shot_opt->getPointToShootAt() - assigned_robot.position();
+            abs_angle_between_pass_and_shot_vectors =
+                (robot_to_ball.orientation() - robot_to_shot_target.orientation())
+                    .clamp()
+                    .abs();
+
+            Angle goal_angle =
+                acuteAngle(world.field().friendlyGoalpostPos(), assigned_robot.position(),
+                           world.field().friendlyGoalpostNeg())
+                    .abs();
+            net_percent_open =
+                best_shot_opt->getOpenAngle().toDegrees() / goal_angle.toDegrees();
+        }
+
+        // If we have a shot with a sufficiently large enough opening, and the
+        // deflection angle that is reasonable, we should one-touch kick the ball
+        // towards the enemy net
+        if (best_shot_opt && net_percent_open > MIN_SHOT_NET_PERCENT_OPEN &&
+            abs_angle_between_pass_and_shot_vectors < MAX_DEFLECTION_FOR_ONE_TOUCH_SHOT)
+        {
+            return best_shot_opt;
+        }
```

Comment:
this can be put in the if statement above so we don't have random `        double net_percent_open;` and `Angle abs_angle_between_pass_and_shot_vectors;` hanging around

Reply:
cleaned up